### PR TITLE
[Pivot Relations] Pivot Selector should not allow to re-select the same.

### PR DESCRIPTION
**Pull request by pxpm**
### BEFORE - What was wrong? What was happening before this PR?

In relations that work with the pivot field `BelongsToMany/MorphToMany` the user could select twice the same pivot entry without noticing and only the last choise of the same entry would get persisted in the database. This could have disastrous consequences if allowed to happen.

### AFTER - What is happening after this PR?

The already selected pivot entries gets disabled in the other pivot selectors preventing them from beeing selected twice.

### How did you achieve that, in technical terms?

Using JS :| Not a fan, but it works. 

When the field initialize we check for other pivots and get all their values, then we disable those options on all other selects.
When value change we re-enable the "free" option and disable the selected one.

### Is it a breaking change or non-breaking change?
Noo

### How can we test the before & after?

Just see `Dummyproducts` in Demo before & after this PR.

### CODE REVIEW

**Comment by pxpm** [comment]
Hey @tabacitu I think I'm kinda happy on the state this is atm, would you mind taking a look ?

File: src/resources/views/crud/fields/relationship/fetch.blade.php [hunk 1]
```diff
@@ -191,16 +196,37 @@ function bpFieldInitFetchElement(element) {
                     },
                     processResults: function (data, params) {
                         params.page = params.page || 1;
+                        
+                        // if field is a pivot selector we are gona get other pivot values so we can disable them from selection.
```
Comment (pxpm) [inline src/resources/views/crud/fields/relationship/fetch.blade.php:200]:
```suggestion
                        // if field is a pivot select we are gona get other pivot values so we can disable them from selection.
```

File: src/resources/views/crud/fields/relationship/fetch_or_create.blade.php [hunk 2]
```diff
@@ -50,46 +50,48 @@
     // this is the time we wait before send the query to the search endpoint, after the user as stopped typing.
     $field['delay'] = $field['delay'] ?? 500;
 
+    // this field can be used as a pivot selector for n-n relationships
```
Comment (pxpm) [inline src/resources/views/crud/fields/relationship/fetch_or_create.blade.php:53]:
```suggestion
    // this field can be used as a pivot select for n-n relationships
```

File: src/resources/views/crud/fields/relationship/fetch_or_create.blade.php [hunk 3]
```diff
@@ -539,16 +543,36 @@ function bpFieldInitFetchOrCreateElement(element) {
             },
             processResults: function (data, params) {
                 params.page = params.page || 1;
+
+                // if field is a pivot selector we are gona get other pivot values so we can disable them from selection.
```
Comment (pxpm) [inline src/resources/views/crud/fields/relationship/fetch_or_create.blade.php:547]:
```suggestion
                // if field is a pivot select we are gona get other pivot values so we can disable them from selection.
```

File: src/resources/views/crud/fields/relationship/relationship_select.blade.php [hunk 4]
```diff
@@ -7,6 +7,9 @@
     $field['allows_null'] = $field['allows_null'] ?? $crud->model::isColumnNullable($field['name']);
     // Note: isColumnNullable returns true if column is nullable in database, also true if column does not exist.
 
+    // this field can be used as a pivot selector for n-n relationships
```
Comment (pxpm) [inline src/resources/views/crud/fields/relationship/relationship_select.blade.php:10]:
```suggestion
    // this field can be used as a pivot select for n-n relationships
```

**Comment by tabacitu** [comment]
Thank you @pxpm . One thing that triggers my spidey-sense is that... it's invasive. We're making changes inside `fetch`, `fetch_or_create` and `relationship_select`... that are ONLY needed if they're loaded by `repeatable_relation`. Or if you look at it the other way around: for a feature offered by `repeatable_relation`... we're bloating up 3 other subfields.

I guess I didn't see this before, because the changes where a lot smaller (to 2 files if I remember correctly).

---

Let's do this, please. Could you please create **a separate PR**, which would have 90-100% of all changes inside `repeatable_relation`? It _should_ be possible. If not with this functionality (disabled options), then with the other one we talked about (don't let it select and show bubble notification). We should have MOST if not ALL of needed Javascript inside `repeatable_relation`:
- it's his functionality, after all;
- inside `repeatable_relation` we already know what the pivot select is, and what _type_ of field; 
- we should be able to add JS at the bottom of `repeatable_relation` that _overrides_ the JS in the subfield;

I think we should limit ourselves at what we can do from within `repeatable_relation`. So yes, the solution is JS. But JS inside `repeatable_relation`, not JS inside all fields that could be used as the pivot_select inside `repeatable_relation`.

Don't you think that would make these a lot easier to maintain?

**Comment by pxpm** [comment]
Hey @tabacitu I've tried to give this a go, and I think it will be very difficult to make it work this way using JS inside `repeatable_relation`. 

It messes with AJAX, it messes with field events, `repeatable_relation` is not a "field", so it doesn't even have a `initFunction` and will never have, because the field loaded in the page is the `repeatable`, so atmost we could add this functionality to repeatable. 

I still think you are not looking at from the rigth perspective. "Technically" the change is only in one field (a select2). It turns out to be in three because we have the same field split by functionality and not by field type. Any of these three fields (select, fetch and fetch_and_create) is a "select" per-se. If we had everything coupled inside of the same file, this change would not require that many files. 🙃 

Also there are other functionality that we have in fields like these, for example the dependant functionality, that I guess 95% of time is disabled, but when needed ... We also have the `processItemText` that is shared between those files, just because the field type is split by functionality and not by type like I previously said. 

If we kept our old "mentality" probably we would create a "pivot_select" field and just cram everything inside that file, then when some change is needed in the "select2" we would endup changing 4 files instead of 3, just because we created another select2 to split functionality, but the field stills the same.

Let me know what you think, I freaking hate the other solution, I would never do it like that for myselft so I am very reluctant to do it for the others.

**Comment by tabacitu** [comment]
Tested, works like a charm, merged 🎉

**Pull request merged**